File: reference/burn_trainer_reference.py
# /// script
# requires-python = ">=3.10"
# dependencies = ["torch==2.*", "numpy"]
# ///
"""PyTorch reference for the `BurnTrainer` step-loss golden (#49 H9).

The convergence tests only assert a *trend* (`last < 0.7 * first`), which a
trainer that "trains fast but wrong" passes. This script pins the trainer's
exact per-step losses to an INDEPENDENT PyTorch computation, the way
`lora_reference.py` pins the toy forward pass.

## Why this one needs a dump

Every other reference here derives its inputs from fixed constants, so torch and
burn start from identical numbers by construction. `BurnTrainer`'s synthetic run
cannot: its frozen base, its LoRA `A` init, and its Gaussian-blob dataset all
come from burn's seeded `StdRng` (ChaCha), which PyTorch cannot reproduce. So we
feed torch burn's *actual* initial tensors and data — dumped by
`cargo run -p loractl-core --example dump_synthetic_run` — and have torch
independently recompute the **losses** from them.

What that does and does not prove:

  * INDEPENDENT: the forward (frozen `fc1` -> ReLU -> base + (alpha/rank)·B(A(h))),
    the cross-entropy, the AdamW update, the freeze (only A/B are optimized), and
    the record-loss-before-step ordering are all recomputed by torch. A bug in any
    of them shows up as a loss mismatch. This is a real numerics proof.
  * SHARED: the initial weights and the training data come from burn. A bug in
    burn's *data generation* (e.g. centroids drawn with the wrong scale) is not
    caught here — it is a shared input to both sides. That half is covered by the
    black-box convergence test, which would stop converging.

## AdamW: burn == torch, on purpose

burn's `AdamW` (burn-optim 0.21 `adamw.rs`) computes
`p := p*(1 - lr*wd) - lr * m̂/(sqrt(v̂) + eps)` with defaults
`beta_1=0.9, beta_2=0.999, epsilon=1e-5` — i.e. decoupled decay, identical in
form to `torch.optim.AdamW`. torch's eps default is 1e-8, so it is set to 1e-5
explicitly below. The golden pins two trajectories, `weight_decay = 0.0` and
`0.05`, so the *decoupled* decay semantics are pinned against torch and not just
asserted (see `.claude/rules/burn-optimizer-and-dropout.md`).

Convention: all matrices in the dump are burn layout `[d_in, d_out]`. This script
owns every transpose (torch `nn.Linear.weight` is `[out, in]`) and transposes back
before dumping, so the Rust test does ZERO transposing.
"""

import argparse
import json
import platform
import sys
from pathlib import Path

import numpy as np
import torch
import torch.nn.functional as F

torch.set_default_dtype(torch.float32)

# burn-optim 0.21 AdamWConfig defaults (adamw.rs) — NOT torch's defaults.
BETAS = (0.9, 0.999)
EPS = 1e-5


def load(dump: Path, name: str, shapes: dict, dtype: str) -> np.ndarray:
    suffix = "f32" if dtype == "f32" else "i64"
    raw = np.fromfile(dump / f"{name}.{suffix}.bin", dtype="<f4" if dtype == "f32" else "<i8")
    return raw.reshape(shapes[name])


def build(dump: Path, manifest: dict):
    shapes = manifest["shapes"]
    hp = manifest["hyperparams"]
    rank = int(hp["rank"])

    fc1_w = load(dump, "fc1_weight", shapes, "f32")  # [784, 256] burn layout
    fc1_b = load(dump, "fc1_bias", shapes, "f32")  # [256]
    base_w = load(dump, "fc2_base_weight", shapes, "f32")  # [256, 10]
    base_b = load(dump, "fc2_base_bias", shapes, "f32")  # [10]
    a_w = load(dump, "lora_a_weight", shapes, "f32")  # [256, rank]
    b_w = load(dump, "lora_b_weight", shapes, "f32")  # [rank, 10]

    d_in, hidden = fc1_w.shape
    out = base_w.shape[1]

    fc1 = torch.nn.Linear(d_in, hidden, bias=True)
    fc1.weight.data = torch.tensor(fc1_w.T.copy())
    fc1.bias.data = torch.tensor(fc1_b.copy())
    fc1.weight.requires_grad_(False)  # FROZEN — excluded from the optimizer
    fc1.bias.requires_grad_(False)

    base = torch.nn.Linear(hidden, out, bias=True)
    base.weight.data = torch.tensor(base_w.T.copy())
    base.bias.data = torch.tensor(base_b.copy())
    base.weight.requires_grad_(False)  # FROZEN
    base.bias.requires_grad_(False)

    lora_a = torch.nn.Linear(hidden, rank, bias=False)
    lora_a.weight.data = torch.tensor(a_w.T.copy())
    lora_b = torch.nn.Linear(rank, out, bias=False)
    lora_b.weight.data = torch.tensor(b_w.T.copy())

    batches = []
    for i in range(int(manifest["batches_dumped"])):
        x = torch.tensor(load(dump, f"batch_{i}_x", shapes, "f32").copy())
        y = torch.tensor(load(dump, f"batch_{i}_y", shapes, "i64").copy())
        batches.append((x, y))

    return fc1, base, lora_a, lora_b, batches


def run(dump: Path, manifest: dict, weight_decay: float):
    """Replay `run_classification`'s loop in torch and return its losses."""
    hp = manifest["hyperparams"]
    steps, lr, scaling = int(hp["steps"]), float(hp["lr"]), float(hp["scaling"])
    fc1, base, lora_a, lora_b, batches = build(dump, manifest)

    opt = torch.optim.AdamW(
        [lora_a.weight, lora_b.weight],
        lr=lr,
        betas=BETAS,
        eps=EPS,
        weight_decay=weight_decay,
        amsgrad=False,
    )

    losses = []
    for step in range(steps):
        x, y = batches[step % len(batches)]
        h = torch.relu(fc1(x))
        # LoRA forward: base(h) + (alpha/rank) · B(A(h)). Dropout is 0.0, and
        # burn's Dropout is identity at prob 0, so there is nothing to model.
        logits = base(h) + scaling * lora_b(lora_a(h))
        # burn's CrossEntropyLoss (no weights, no smoothing) is
        # `-mean(log_softmax(logits)[target])` == torch's mean-reduction CE.
        loss = F.cross_entropy(logits, y)
        losses.append(float(loss.item()))  # record BEFORE backward/step
        opt.zero_grad()
        loss.backward()
        opt.step()

    # Frozen base must be untouched — the strongest form of the LoRA claim.
    assert fc1.weight.grad is None and base.weight.grad is None

    # Back to burn layout `[d_in, d_out]` before dumping.
    b_final = lora_b.weight.detach().cpu().numpy().T.copy()
    return losses, b_final


parser = argparse.ArgumentParser()
parser.add_argument("--dump", required=True, type=Path, help="dir written by dump_synthetic_run")
args = parser.parse_args()

manifest = json.loads((args.dump / "manifest.json").read_text())
hp = manifest["hyperparams"]

trajectories = {}
for wd in hp["weight_decays"]:
    losses, b_final = run(args.dump, manifest, float(wd))
    trajectories[f"wd_{wd}"] = {
        "weight_decay": float(wd),
        "losses": losses,
        "lora_b_final": b_final.flatten().tolist(),
    }

print("=== BurnTrainer step-loss reference (torch replay of burn's init + data) ===", file=sys.stderr)
print(f"torch: {torch.__version__}  platform: {platform.platform()}", file=sys.stderr)
for key, t in trajectories.items():
    ls = t["losses"]
    print(f"{key}: {ls[0]:.6f} -> {ls[-1]:.6f} over {len(ls)} steps", file=sys.stderr)
a, b = (trajectories[f"wd_{wd}"]["losses"] for wd in hp["weight_decays"])
print(f"max |loss diff| between the two decay settings: {max(abs(x - y) for x, y in zip(a, b)):.3e}", file=sys.stderr)

golden = {
    "provenance": {
        "torch_version": torch.__version__,
        "platform": platform.platform(),
        "note": "burn-layout [d_in,d_out]; regenerate with `just burn-trainer-reference`",
    },
    "hyperparams": {
        "seed": int(hp["seed"]),
        "steps": int(hp["steps"]),
        "rank": int(hp["rank"]),
        "alpha": float(hp["alpha"]),
        "lr": float(hp["lr"]),
        "scaling": float(hp["scaling"]),
    },
    "trajectories": [trajectories[f"wd_{wd}"] for wd in hp["weight_decays"]],
}
print(json.dumps(golden, indent=2))

File: crates/loractl-core/tests/burn_trainer_reference.rs
//! `BurnTrainer` step-loss golden vs PyTorch (#49 H9).
//!
//! `convergence.rs` only asserts a *trend* (`last < 0.7 * first`) — a trainer
//! that trains fast but **wrong** (a mis-scaled LoRA delta, a dropped
//! `weight_decay`, a coupled-instead-of-decoupled decay, a loss read after the
//! optimizer step) sails through it. This test pins the *exact* per-step loss
//! the public `Trainer` emits, against a checked-in golden computed
//! independently by PyTorch (`reference/burn_trainer_reference.py`), the way
//! `lora_reference.rs` pins the toy forward pass.
//!
//! ## What is and isn't independent
//!
//! burn's frozen base, LoRA `A` init, and Gaussian-blob dataset all come from
//! its seeded ChaCha `StdRng`, which PyTorch cannot reproduce — so the golden
//! cannot be derived end-to-end in Python. Instead, `just
//! burn-trainer-reference` dumps burn's **actual** init + batches
//! (`examples/dump_synthetic_run.rs`, via `burn_trainer::synthetic_run_inputs`)
//! and torch recomputes the **losses** from them:
//!
//! * **Independent** (this is the numerics proof): the forward pass, the
//!   `alpha/rank` scaling, the cross-entropy, the AdamW update incl. decoupled
//!   weight decay, the freeze, and the record-loss-*before*-step ordering.
//! * **Shared**: the initial weights and the training data. A bug in burn's
//!   *data generation* is a shared input and is not caught here — the black-box
//!   `convergence.rs` covers that half (it would stop converging).
//!
//! The dump path is itself gated by this test: `synthetic_run_inputs` has to
//! reproduce `run_classification`'s init and RNG ordering exactly (see
//! `.claude/rules/burn-lazy-param-init.md`) or torch is fed weights the trainer
//! never had and every loss below mismatches.
//!
//! Always-run, offline, no Python at `cargo test` time. Regenerate with
//! `just burn-trainer-reference` when the trainer's math, burn's version, or the
//! constants below change.

use burn::tensor::{TensorData, Tolerance};
use loractl_core::adapter::load_adapter;
use loractl_core::config::{
    ComputeConfig, DatasetConfig, FlowConfig, LoraConfig, ModelConfig, OptimConfig, OutputConfig,
    TaskKind,
};
use loractl_core::{BurnTrainer, Device, LoraMlp, NdArray, TrainConfig, TrainEvent, Trainer};
use std::path::{Path, PathBuf};

// The run under test. MUST match `examples/dump_synthetic_run.rs` — and the
// `hyperparams` assertions below mechanically enforce it: the golden records the
// values the dump was taken with, so a drift between the two fails loudly here
// instead of silently pinning a different run.
const SEED: u64 = 7;
const STEPS: u64 = 12;
const RANK: u32 = 4;
const ALPHA: f32 = 8.0;
const LR: f64 = 0.01;
/// `NUM_CLASSES` in `burn_trainer.rs` — `lora_b` is `[rank, out]`.
const NUM_CLASSES: usize = 10;

/// Tolerance: burn (ndarray) and torch (Accelerate) do the same arithmetic in a
/// different summation order over 784-wide dots, and the difference compounds
/// across 12 AdamW steps. Measured max |diff| on this machine: ~1e-5 (the suite
/// still passes at `absolute(1e-4)`), so 1e-3 leaves ~100x headroom for a CI
/// host with a different BLAS while staying a real gate — the mutants this test
/// is aimed at move the losses by 1e-2 or more (dropping `weight_decay` alone
/// separates the trajectories by 5e-2).
fn tol() -> Tolerance<f32> {
    Tolerance::absolute(1e-3)
}

#[derive(serde::Deserialize)]
struct Golden {
    hyperparams: Hyperparams,
    trajectories: Vec<Trajectory>,
}

#[derive(serde::Deserialize)]
struct Hyperparams {
    seed: u64,
    steps: u64,
    rank: u32,
    alpha: f32,
    lr: f64,
}

#[derive(serde::Deserialize)]
struct Trajectory {
    weight_decay: f64,
    losses: Vec<f32>,
    lora_b_final: Vec<f32>,
}

/// A unique temp output dir, removed on drop.
struct TempDir(PathBuf);

impl TempDir {
    fn new(tag: &str) -> Self {
        let nanos = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let dir =
            std::env::temp_dir().join(format!("loractl-{tag}-{}-{nanos}", std::process::id()));
        Self(dir)
    }
}

impl Drop for TempDir {
    fn drop(&mut self) {
        let _ = std::fs::remove_dir_all(&self.0);
    }
}

fn config(out_dir: &Path, weight_decay: f64) -> TrainConfig {
    TrainConfig {
        steps: STEPS,
        seed: SEED,
        task: TaskKind::Classification,
        model: ModelConfig {
            base: "synthetic".into(),
        },
        lora: LoraConfig {
            rank: RANK,
            alpha: ALPHA,
            // Non-zero dropout would draw RNG inside the training loop, which
            // torch cannot replay — the golden run is dropout-free by design.
            dropout: 0.0,
            targets: vec![],
        },
        dataset: DatasetConfig {
            path: PathBuf::from("unused"),
            resolution: 28,
        },
        optim: OptimConfig {
            lr: LR,
            weight_decay,
        },
        output: OutputConfig {
            dir: out_dir.to_path_buf(),
            name: "adapter".into(),
            checkpoint_every: u64::MAX,
            sample_every: 0,
        },
        compute: ComputeConfig::default(),
        flow: FlowConfig::default(),
    }
}

#[test]
fn burn_trainer_step_losses_match_pytorch_golden() {
    let golden: Golden = serde_json::from_str(include_str!("golden/burn_trainer_steps.json"))
        .expect("golden fixture parses");

    // Drift guard: the golden is only meaningful for the run it was generated
    // from. If `examples/dump_synthetic_run.rs` and this file disagree about the
    // run, fail here — not with a confusing numerics mismatch 40 lines down.
    let hp = &golden.hyperparams;
    assert_eq!(
        (hp.seed, hp.steps, hp.rank, hp.alpha, hp.lr),
        (SEED, STEPS, RANK, ALPHA, LR),
        "the golden was generated for a different run — regenerate with \
         `just burn-trainer-reference` after changing the constants"
    );
    assert_eq!(
        golden.trajectories.len(),
        2,
        "the golden pins two weight-decay trajectories (0.0 and the kill-test value)"
    );

    let device: Device<NdArray> = Default::default();

    for (i, traj) in golden.trajectories.iter().enumerate() {
        let out = TempDir::new(&format!("h9-{i}"));
        let config = config(&out.0, traj.weight_decay);

        // Black-box: observe ONLY the public event stream, exactly as the CLI
        // and the API do.
        let mut losses = Vec::new();
        let mut trainer = BurnTrainer;
        let adapter = trainer
            .train(&config, &mut |event| {
                if let TrainEvent::Step { loss, .. } = event {
                    losses.push(loss);
                }
            })
            .expect("the training run succeeds");

        assert_eq!(
            losses.len(),
            STEPS as usize,
            "exactly one Step event per step"
        );
        TensorData::new(losses, [STEPS as usize]).assert_approx_eq::<f32>(
            &TensorData::new(traj.losses.clone(), [STEPS as usize]),
            tol(),
        );

        // The final trained `lora_b`, read back through the public adapter path.
        // It is zero-initialized, so it carries the ENTIRE learned update — and
        // pinning it catches an error in the last optimizer step, which no
        // recorded loss ever observes.
        let reloaded: LoraMlp<NdArray> =
            load_adapter(&adapter, &device).expect("the final adapter loads");
        reloaded
            .fc2
            .lora_b
            .weight
            .val()
            .into_data()
            .assert_approx_eq::<f32>(
                &TensorData::new(traj.lora_b_final.clone(), [RANK as usize, NUM_CLASSES]),
                tol(),
            );
    }
}
